mirai-core/src/commonMain/kotlin/message/MultiMsgUploader.kt

```diff
@@ -0,0 +1,250 @@
+            // Upload nested and refine to service msg
+            val nestedMMUploader = newUploader()
+            nestedMMUploader.emitMain(nestedForward.nodeList)
+
+            val resId = nestedMMUploader.uploadAndReturnResId()
+
+            val mirror = nestedMMUploader.nestedMsgs
+            mirror.remove("MultiMsg")
+            nestedMsgs.putAll(mirror)
+
+            val nid = newNid()
+            nestedMsgs[nid] = nestedMMUploader.mainMsg
+            return messageChainOf(
+                RichMessage.forwardMessage(
+                    resId = resId,
+                    fileName = nid,
+                    forwardMessage = nestedForward,
+                )
+            )
+        }
+
+        msgChain.firstIsInstanceOrNull<MessageOrigin>()?.let { origin ->
+            convertByMessageOrigin(origin)?.let { return it }
+        }
+
+        return convertByReUpload()
+    }
+
+    open suspend fun emit(id: String, msgs: Collection<ForwardMessage.INode>) {
+        val nds = mutableListOf<MsgComm.Msg>().let { tmp ->
+            nestedMsgs.putIfAbsent(id, tmp) ?: tmp
+        }
+
+        val existsIds = mutableSetOf<Long>()
+
+        msgs.forEach { msg ->
+            var msgChain = msg.messageChain
+            msgChain.takeSingleContent<ForwardMessage>()?.let { nestedForward ->
+                msgChain = convertNestedForwardMessage(nestedForward, msgChain)
+            }
+
+            msgChain = handler.conversionMessageChain(msgChain)
+
+            var seq: Int = -1
+            var uid: Int = -1
+            msg.messageChain.sourceOrNull?.let { source ->
+                source as MessageSourceInternal
+
+                seq = source.sequenceIds.first()
+                uid = source.internalIds.first()
+            }
+            while (true) {
+                if (seq != -1 && uid != -1) {
+                    if (existsIds.add(seq.concatAsLong(uid))) break
+                }
+                seq = tmpRand.nextInt().absoluteValue
+                uid = tmpRand.nextInt().absoluteValue
+            }
+
+            val msg0 = MsgComm.Msg(
+                msgHead = MsgComm.MsgHead(
+                    fromUin = msg.senderId,
+                    toUin = if (isLong) {
+                        handler.targetUserUin ?: 0
+                    } else 0,
+                    msgSeq = seq,
+                    msgTime = msg.time,
+                    msgUid = 0x01000000000000000L or uid.toLongUnsigned(),
+                    mutiltransHead = MsgComm.MutilTransHead(
+                        status = 0,
+                        msgId = 1,
+                    ),
+                    msgType = 82, // troop,
+                    groupInfo = handler.run { msg.groupInfo },
+                    isSrcMsg = false,
+                ),
+                msgBody = ImMsgBody.MsgBody(
+                    richText = ImMsgBody.RichText(
+                        elems = msgChain.toMessageChain()
```

Comment:
```suggestion
                        elems = msgChain
```
redundant? (check before applying suggestion)